File: bench/matmul/matmul.js
// Dense matrix multiplication: nested loops over indexed 2-D arrays.
const n = 240;

const a = [];
const b = [];
for (let i = 0; i < n; i++) {
  const ra = new Array(n);
  const rb = new Array(n);
  for (let j = 0; j < n; j++) {
    ra[j] = i + j;
    rb[j] = i - j;
  }
  a.push(ra);
  b.push(rb);
}

let total = 0;
for (let i = 0; i < n; i++) {
  const ai = a[i];
  for (let j = 0; j < n; j++) {
    let sum = 0;
    for (let k = 0; k < n; k++) {
      sum += ai[k] * b[k][j];
    }
    total += sum;
  }
}

console.log(total);

File: bench/matmul/matmul.lua
-- Dense matrix multiplication: nested loops over indexed 2-D tables.
local n = 240

local a = {}
local b = {}
for i = 0, n - 1 do
    local ra = {}
    local rb = {}
    for j = 0, n - 1 do
        ra[j] = i + j
        rb[j] = i - j
    end
    a[i] = ra
    b[i] = rb
end

local total = 0
for i = 0, n - 1 do
    local ai = a[i]
    for j = 0, n - 1 do
        local sum = 0
        for k = 0, n - 1 do
            sum = sum + ai[k] * b[k][j]
        end
        total = total + sum
    end
end

print(total)
